ArtOfIllusion/src/artofillusion/view/ClickedPointFinder.java

```diff
@@ -0,0 +1,186 @@
+		Vec3 cameraZ = v.getCamera().getCameraCoordinates().getZDirection();
+
+		Mat4 objToScreen = v.getCamera().getObjectToScreen();
+		Mat4 worldToScreen = v.getCamera().getWorldToScreen();
+		Mat4 toScene, toThisObject, fromExtToLocal;
+		ObjectInfo[] objList = renderableObjects(v);
+		RenderingMesh rMesh;
+		Vec3[] corner3D = new Vec3[3];
+		Vec2[] corner2D = new Vec2[3];
+		Vec2 point2D = new Vec2(point.x, point.y);
+		CoordinateSystem localCoords;
+
+		localCoords = getLocalCoords(v);
+		toThisObject = localCoords.toLocal();
+
+		for (int i = 0; i< objList.length ; i++)
+		{	
+			rMesh = objList[i].getPreviewMesh();
+			toScene = objList[i].getCoords().fromLocal();
+			fromExtToLocal = toThisObject.times(toScene);
+
+			for (int j = 0; j < rMesh.triangle.length; j++)
+			{
+				corner3D[0] = new Vec3 (fromExtToLocal.times(rMesh.vert[rMesh.triangle[j].v1]));
+				corner3D[1] = new Vec3 (fromExtToLocal.times(rMesh.vert[rMesh.triangle[j].v2]));
+				corner3D[2] = new Vec3 (fromExtToLocal.times(rMesh.vert[rMesh.triangle[j].v3]));				
+
+				corner2D[0] = worldToScreen.timesXY(corner3D[0]);
+				corner2D[1] = worldToScreen.timesXY(corner3D[1]);
+				corner2D[2] = worldToScreen.timesXY(corner3D[2]);
+				
+				if (v instanceof ObjectViewer && ! ((ObjectViewer)v).getUseWorldCoords())
+				{
+					corner2D[0] = objToScreen.timesXY(corner3D[0]);
+					corner2D[1] = objToScreen.timesXY(corner3D[1]);
+					corner2D[2] = objToScreen.timesXY(corner3D[2]);
+				}
+				if (onTriangle(corner2D, point2D))
+				{	
+					(corner3D[0] = corner3D[0]).scale(bary[0]);
+					(corner3D[1] = corner3D[1]).scale(bary[1]);
+					(corner3D[2] = corner3D[2]).scale(bary[2]);
+				
+					pointOnTriangle = new Vec3(corner3D[0].plus(corner3D[1].plus(corner3D[2])));
+					if (inSpace)
+					{	
+						clickedPoint = pointOnTriangle;
+						inSpace = false;
+					}
+					else
+						if (closer(pointOnTriangle, clickedPoint, cameraOrigin, cameraZ) && onFrontSide(pointOnTriangle, cameraOrigin, cameraZ))
+							clickedPoint = pointOnTriangle;
+				}
+			}
+		}
+		return clickedPoint;
+	}
+
+	private boolean closer(Vec3 p1, Vec3 p2, Vec3 cameraOrigin, Vec3 cameraZ)
+	{
+		double distTo1 = p1.minus(cameraOrigin).dot(cameraZ);
+		double distTo2 = p2.minus(cameraOrigin).dot(cameraZ);
+
+		System.out.println("distTo1 = " + distTo1);
+		System.out.println("distTo2 = " + distTo2);
+		
+		if (distTo1 < distTo2)
+			return true;
+		else
+			return false;
+	}
+
+	private boolean onFrontSide(Vec3 p, Vec3 cameraOrigin, Vec3 cameraZ)
+	{
+		double distToP = p.minus(cameraOrigin).dot(cameraZ);
+
+		System.out.println("distToP = " + distToP);
+		
+		if (distToP > 0.0)
+			return true;
+		else
+			return false;
+	}
+
+	private boolean onTriangle(Vec2[] corner2D, Vec2 point2D)
+	{
+		// The sum of barycentric coordinates is always 1.0. 
+		// If the point is outside, some of those will be negative. 
+		bary = T.bary(corner2D, point2D);		
```

Comment:
And here, instead of using the object reference `T.` you would call `bary = TriangleMath.bary(corner2d, point2d);`

Or, if you are using a lot of TriangleMath, you could add `TriangleMath.*` to the imports list. Then you would call: `bary = bary(corner2d, point2d);`